src/signalrclient/json_helpers.cpp

```diff
@@ -57,7 +58,42 @@ namespace signalr
             // because the server expects certain values to be 1 instead of 1.0 (like protocol version)
             if (std::modf(value, &intPart) == 0)
             {
-                return Json::Value((int)intPart);
+				if (value < 0)
+				{
+					if (value > (double)INT32_MIN)
+					{
+						// Fits within int32_t
+						return Json::Value((int32_t)intPart);
+					}
+					else
+					{
+						// Needs int64_t
+						return Json::Value((int64_t)intPart);
+					}
+				}
+				else
+				{
+					if (value < INT32_MAX)
+					{
+						// Fits within int32_t
+						return Json::Value((int32_t)intPart);
+					}
+					else if (value < UINT32_MAX)
+					{
+						// Needs uint32_t
+						return Json::Value((uint32_t)intPart);
+					}
+					else if (value < INT64_MAX)
+					{
+						// Fits within int64_t
+						return Json::Value((int64_t)intPart);
+					}
+					else
+					{
+						// Needs uint64_t
+						return Json::Value((uint64_t)intPart);
```

Comment:
Given that [doubles can contain exponents](https://en.wikipedia.org/wiki/IEEE_754), I'm pretty sure a non-fractional double can represent numbers larger than uint64_t's max value. The same issue exists for "large" negative numbers. For these values, we should just leave them as doubles.

Reply:
@halter73, thanks for all of the comments.  This is feeling a bit like trial by fire for my first pull request but I'll keep working on it.

I agree with adding in logic to fall back to double for numbers outside of the range of the integer types.  I do have a question regarding supporting all of the integer types, however.  Given that in most cases the server side will be written in C#, does the SignalR server code take care of converting to the type that is given in the server-side methods?  If we were to implement this change to use just int64_t as you suggest, would the SignalR code handle the possible truncation if the server-side method specified its parameter as int or short?  Also, given that int64_t is signed, wouldn't that also open us back up to the overflow error I mentioned in #17779 (understanding that it would be far less likely to happen)?  If we use the smaller integer types on the C++ client, as I have it implemented now, would the SignalR server code handle the conversion if the server-side method specified its parameter as long or int? (I don't think it would be a problem but I figured I should ask)

I'm not sure how to go about adding in tests for this.  There doesn't seem to be an existing source file in the tests project that is really related to this.  Any suggestions you have in where to add the new tests would be greatly appreciated.

Reply:
> Given that in most cases the server side will be written in C#, does the SignalR server code take care of converting to the type that is given in the server-side methods?

I think this is left up to the JSON serializer. Since JSON/JavaScript has only one "number" type which is stored in the IEEE 754 binary64 (i.e. float64_t or double\*) format, there's really no decision for the SignalR server to make regarding what type a number should be converted to.

It is possible that precision might be lost when converting an int64_t to a float64_t, but there should never be any truncation or overflow since the float64_t can use the exponent to represent values both much larger and much smaller than what can be represented using int64_t.

In general, SignalR does not use the payload itself to determine the type of a given number. On the server, if SignalR was expecting a double, it will deserialize the type as a double even if the value is is `0` which could easily be deserialized to a ushort or even a byte. The only reason the C++ client needs to basically guess the type of the number, is because c++ doesn't the powerful reflection as C# has.

I'm starting to think it might be simpler for the SignalR C++ client to treat all numbers as float64_t and let users do any conversion themselves if it's necessary.

> I'm not sure how to go about adding in tests for this. There doesn't seem to be an existing source file in the tests project that is really related to this. Any suggestions you have in where to add the new tests would be greatly appreciated.

@BrennanConroy Do you know of any similar tests?

\* In C++, a double might not actually be 64 bits

Reply:
> On the server, if SignalR was expecting a double, it will deserialize the type as a double even if the value is is `0`

The "problem" is that with System.Text.Json as the default serializer now, it is a lot more strict about converting numbers between double and int, which is the whole reason the `std::modf` code exists.

We shied away from adding support for integral types to the `signalr::value` type until we needed it (like when adding MsgPack), but this conversation is making me think we might want to scrap the workaround and just add first party support for integral types.

Reply:
> Do you know of any similar tests?

I would probably copy https://github.com/aspnet/SignalR-Client-Cpp/blob/0a35552d0361833f938ece2fce45e60c244b935e/test/signalrclienttests/hub_connection_impl_tests.cpp#L410 and modify it for different argument types.

Reply:
> We shied away from adding support for integral types to the signalr::value type until we needed it (like when adding MsgPack), but this conversation is making me think we might want to scrap the workaround and just add first party support for integral types.

The System.Text.Json thing is unfortunate, but I think we should stick with float64 as the only supported numeric type in the C++ client for not. Serializing as int64_t if possible in order to remove the decimal point and staying with float64_t when there's a fractional part or the number's too large seems to be the simplest fix for now. Hopefully we can get System.Text.Json to relax its behavior at some point.

Reply:
> > Do you know of any similar tests?
> 
> I would probably copy
> 
> https://github.com/aspnet/SignalR-Client-Cpp/blob/0a35552d0361833f938ece2fce45e60c244b935e/test/signalrclienttests/hub_connection_impl_tests.cpp#L410
> 
> and modify it for different argument types.

Thanks for the suggestion.  To be forthcoming, C++ is not my primary language and I'm not familiar with the testing framework being used here.  I will try to find some time to develop a test method but it will take some time.

Reply:
> > Do you know of any similar tests?
> 
> I would probably copy
> 
> https://github.com/aspnet/SignalR-Client-Cpp/blob/0a35552d0361833f938ece2fce45e60c244b935e/test/signalrclienttests/hub_connection_impl_tests.cpp#L410
> 
> and modify it for different argument types.

@BrennanConroy, I have been looking over the available test cases in the above and have not seen any examples of sending non-string payloads.  I also tried looking at the other tests but haven't found any non-string examples there either.  I'm not sure if I'm just missing them, however.  I am able to run the tests and have been experimenting with adding my own test cases but I could use some help with sending a non-string payload since the problem I'm trying to address is obviously related to that.  Thanks...

Reply:
> have not seen any examples of sending non-string payloads

You're right, looks like I'll need to add some more tests later.

For now you could try looking at https://github.com/aspnet/SignalR-Client-Cpp/blob/0a35552d0361833f938ece2fce45e60c244b935e/test/signalrclienttests/hub_connection_impl_tests.cpp#L796 and modifying it to send a non-empty array for args.

```c++
std::vector<signalr::value> arr { 100 };
hub_connection->invoke("method", signalr::value(arr), ...);
```

Reply:
Added `TEST(invoke, invoke_handles_UINT32_MAX_value_as_unsigned_int)` test to validate that the issue I originally posted would now be handled correctly.  Probably still need 3-4 more tests to exercise all of the conditions added to the signalr::createJson() method in json_helpers.cpp.

Reply:
I've learned a lot more about the double (aka. float64) type in working on this PR.  All of the tests I have implemented currently pass but I'm beginning to rethink the edge cases because, as it turns out, double only has 52 bits to represent its mantissa (excluding the sign bit) so integral values contained within a double will be truncated if they exceed this.  I made the tests "work" by restricting the level of precision at which the values are compared.

This takes us back to the point that @BrennanConroy made on Dec 11, 2019: 

> > On the server, if SignalR was expecting a double, it will deserialize the type as a double even if the value is is `0`
> 
> The "problem" is that with System.Text.Json as the default serializer now, it is a lot more strict about converting numbers between double and int, which is the whole reason the `std::modf` code exists.
> 
> We shied away from adding support for integral types to the `signalr::value` type until we needed it (like when adding MsgPack), but this conversation is making me think we might want to scrap the workaround and just add first party support for integral types.

This PR, as implemented, will work for integral values that fit within 52 bits which gets us more support than before (i.e. it happens to support the case that I needed, a 32-bit mask ;-)) but it will fail for use cases outside of this range if full fidelity of the integral values are needed (e.g. using a 64-bit mask).  First party support for integral types would seem to be the only way to support those use cases.